sycl/test-e2e/Annotated_usm/annotated_usm.cpp

```diff
@@ -0,0 +1,173 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+// E2E tests for annotated USM allocation functions
+
+#include <sycl/sycl.hpp>
+
+#include <complex>
+#include <numeric>
+
+// clang-format on
+
+using namespace sycl::ext::oneapi::experimental;
+using namespace sycl::ext::intel::experimental;
+using alloc = sycl::usm::alloc;
+
+// Single test instance consisting of (i) calling malloc function `f` (ii)
+// checking usm location of returned pointer  (iii) calling free
+#define TEST_USM_KIND(f, kind, args...)                                        \
+  {                                                                            \
+    auto ap = f(args);                                                         \
+    assert(ap.get() != nullptr &&                                              \
+           sycl::get_pointer_type(ap.get(), Ctx) == kind);                     \
+    free(ap, q);                                                               \
+  }
+
+// Single test instance consisting of (i) calling malloc function `f` (ii)
+// checking alignment Note that the annotated_ptr is not freed in alignment
+// tests to avoid the same base address being returned every time
+#define TEST_ALIGN(f, align, args...)                                          \
+  {                                                                            \
+    auto ap = f(args);                                                         \
+    assert(ap.get() != nullptr && ((uintptr_t)ap.get() % align) == 0);         \
+  }
+
+// Single test instance consisting of (i) calling malloc function `f` (ii)
+// checking return is nullptr (iii) calling free
+#define TEST_ALIGN_NULL_EXPECTED(f, args...)                                   \
+  {                                                                            \
+    auto ap = f(args);                                                         \
+    assert(ap.get() == nullptr);                                               \
+    free(ap, q);                                                               \
+  }
+
+constexpr int N = 10;
+
+template <typename T> void testUsmKind(sycl::queue &q) {
```

Comment:
All of the tests in `testUsmKind(sycl::queue &q)` pass in the queue instead of a device and context. Can we also test the overloads that pass device and context? It would double the number of tests, but that shouldn't be too many!

Also, there might be a better way to organize the tests so we can reduce the number of lines of code.

Reply:
added tests for passing `device` and `context`